### ISSUE TITLE
powershell_out with a code block requires double escape for double quotes

### ISSUE THREAD

**Issue by cordovinian** [chef/chef#5049]
## Description

If I have a script block defined in a library and then pass that to powershell_out only double-escaped double quotes will end up as valid PowerShell syntax. Since the double quotes are missing PowerShell will fail trying to run the code that is no longer in a string.

I think the issue is in the powershell_out mixin where the script block is passed into a command to run PowerShell.

[powershell_out.rb](https://github.com/chef/chef/blob/master/lib/chef/mixin/powershell_out.rb#L94)

_Note_: This is happening inside of a LWRP
## Chef Version

``` powershell
> chef -v
Chef Development Kit Version: 0.14.25
chef-client version: 12.10.24
berks version: 4.3.3
kitchen version: 1.8.0
```
## Platform Version

Windows 10 latest

``` powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      5.0.10586.122
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.10586.122
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```
## Replication Case

``` ruby
   def correct_version_running?
      script_block = <<-EOF
      $myvar= 'variable'
      Write-Verbose 'Only single escape needed for other characters: (\\, \#)' -Verbose
      Write-Verbose \\"String with $myvar interpolation\\" -Verbose
      Write-Verbose \"String with $myvar interpolation\" -Verbose
      Write-Verbose "String with $myvar interpolation" -Verbose
      EOF

      ::Chef::Log.debug("Script block:\n#{script_block}")
      result = powershell_out(script_block)
   end
```
## Client Output

``` cmd
    Mixlib::ShellOut::ShellCommandFailed
    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '1'
    ---- Begin output of powershell.exe -NoLogo -NonInteractive -NoProfile -ExecutionPolicy Unrestricted -InputFormat None -Command "      $myvar= 'variable'
          Write-Verbose 'Only single escape needed for other characters: (\, #)' -Verbose
          Write-Verbose \"String with $myvar interpolation\" -Verbose
          Write-Verbose "String with $myvar interpolation" -Verbose
          Write-Verbose "String with $myvar interpolation" -Verbose
    " ----
    STDOUT: VERBOSE: Only single escape needed for other characters: (\, #)
    VERBOSE: String with variable interpolation
    STDERR: Write-Verbose : A positional parameter cannot be found that accepts argument 'with'.
    At line:4 char:7
    +       Write-Verbose String with $myvar interpolation -Verbose
    +       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : InvalidArgument: (:) [Write-Verbose], ParameterBindingException
        + FullyQualifiedErrorId : PositionalParameterNotFound,Microsoft.PowerShell.Commands.WriteVerboseCommand

    Write-Verbose : A positional parameter cannot be found that accepts argument 'with'.
    At line:5 char:7
    +       Write-Verbose String with $myvar interpolation -Verbose
    +       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : InvalidArgument: (:) [Write-Verbose], ParameterBindingException
        + FullyQualifiedErrorId : PositionalParameterNotFound,Microsoft.PowerShell.Commands.WriteVerboseCommand
    ---- End output of powershell.exe -NoLogo -NonInteractive -NoProfile -ExecutionPolicy Unrestricted -InputFormat None -Command "      $myvar= 'variable'
          Write-Verbose 'Only single escape needed for other characters: (\, #)' -Verbose
          Write-Verbose \"String with $myvar interpolation\" -Verbose
          Write-Verbose "String with $myvar interpolation" -Verbose
          Write-Verbose "String with $myvar interpolation" -Verbose
    " ----
    Ran powershell.exe -NoLogo -NonInteractive -NoProfile -ExecutionPolicy Unrestricted -InputFormat None -Command "      $myvar= 'variable'
          Write-Verbose 'Only single escape needed for other characters: (\, #)' -Verbose
          Write-Verbose \"String with $myvar interpolation\" -Verbose
          Write-Verbose "String with $myvar interpolation" -Verbose
          Write-Verbose "String with $myvar interpolation" -Verbose
    " returned 1
```
## Stacktrace

[chef-stacktrace.txt](https://github.com/chef/chef/files/328952/chef-stacktrace.txt)

**Comment by lamont-granquist** [comment]
The first layer of quoting seems to be due to the heredoc.  If you use `<<'EOF'` that will disable interpolation in the heredoc and will eliminate that layer.

The second layer of quoting comes from the fact that the command is passed to powershell.exe on the command-line inside of double-quotes.

About all I can think of here is that we could render it into a tempfile and pass that to powershell.exe instead in order to get the code off of the command line?

**Comment by lamont-granquist** [comment]
We could go down the road of trying to `.gsub('"', '\"')` but you're going to find so many edge cases with that you'll most likely wind up writing a powershell parser...

**Comment by mwrock** [comment]
One thing here worth noting for anyone looking to fix, `-EncodedCommand` seems like it would be the route for fixing this but allas it is not. With `-EncodedCommand` one could simply base64 the command. However, that would break `powershell_script` on Windows nano which does not support EncodedCommand.

I do think it would be feasible to do a:
```
.gsub('"', "\"")
```
The problem here is that the code block is being injected into another string:

```
 "powershell.exe #{flags.join(' ')} -Command \"#{script}\""
```
Note the quotes around `#{script}` and I think the substitution would cover the vast majority of problem cases.

**Comment by NimishaS** [comment]
Created https://github.com/chef/chef/pull/5381 to fix this issue.

*Note*: `.gsub('"', '\"')` takes care of only one scenario i.e. when double quotes are passed with single escape `(\")`. This seems like a consistent behavior since other special character`(#,/)` also need single escape.
This will not take care of `\\"` and `"`. If we decide to add support for these inputs too, then ideally it should work for all other special characters too.

**Comment by mwrock** [comment]
I think its fine to let the other escape scenarios slide and require users to escape appropriately. What I find particularly aggregious in this double quote scenario is that it is not clear to recipie authors that their code string is being injected into another code string inside of double quotes which will make double quote usage especially problematic.

The beginning of this discource thread is a good example of the confusion I hope such a fix will alleviate: https://discourse.chef.io/t/how-use-stdout-variable-of-mixlib-shellout/9458

**Comment by btm** [comment]
This was fixed by #5430 and released in 12.16.10+
